plugins/modules/proxysql_query_rules_fast_routing.py

```diff
@@ -0,0 +1,470 @@
+        return rule
+
+    def create_rule_config(self, cursor):
+        query_string = \
+            """INSERT INTO mysql_query_rules_fast_routing ("""
+
+        cols = 0
+        query_data = []
+
+        for col, val in iteritems(self.config_data):
+            if val is not None:
+                cols += 1
+                query_data.append(val)
+                query_string += "\n" + col + ","
+
+        query_string = query_string[:-1]
+
+        query_string += (
+            ")\n" +
+            "VALUES (" +
+            "%s ," * cols
+        )
+
+        query_string = query_string[:-2]
+        query_string += ")"
+
+        cursor.execute(query_string, query_data)
+        return True
+
+    def update_rule_config(self, cursor):
+        query_string = "UPDATE mysql_query_rules_fast_routing"
+
+        cols = 0
+        query_data = [
+            self.config_data["username"],
+            self.config_data["schemaname"],
+            self.config_data["flagIN"]
+        ]
+
+        for col, val in iteritems(self.config_data):
+            if val is not None and col not in ("username", "schemaname", "flagIN"):
+                cols += 1
+                query_data.append(val)
+                if cols == 1:
+                    query_string += " SET " + col + "= %s,"
+                else:
+                    query_string += " " + col + " = %s,"
+
+        query_string = query_string[:-1]
+        query_string += (
+            "WHERE username = %s "
+            "AND schemaname = %s "
+            "AND flagIN = %s"
+        )
+
+        cursor.execute(query_string, query_data)
+        return True
+
+    def delete_rule_config(self, cursor):
+        query_string = "DELETE FROM mysql_query_rules_fast_routing"
+
+        cols = 0
+        query_data = []
+
+        for col, val in iteritems(self.config_data):
+            if val is not None:
+                cols += 1
+                query_data.append(val)
+                if cols == 1:
+                    query_string += " WHERE " + col + " = %s"
+                else:
+                    query_string += " AND " + col + " = %s"
+
+        if cols > 0:
+            cursor.execute(query_string, query_data)
+        else:
+            cursor.execute(query_string)
+        check_count = cursor.rowcount
+        return True, int(check_count)
+
+    def manage_config(self, cursor, state):
+        if state:
+            if self.save_to_disk:
+                save_config_to_disk(cursor)
+            if self.load_to_runtime:
+                load_config_to_runtime(cursor)
+
+    def create_rule(self, check_mode, result, cursor):
+        if not check_mode:
+            result['changed'] = self.create_rule_config(cursor)
+            result['msg'] = "Added rule to mysql_query_rules_fast_routing"
+            self.manage_config(cursor, result['changed'])
+            result['rules'] = self.get_rule_config(cursor)
+        else:
+            result['changed'] = True
```

Comment:
When I read right, the code already check existing state in L396 and L399.  
So the `create_rule()` and `update_rule()` functions are only triggered if necessary and their output should always results in `result['changed'] = True`.  
Imho its corrected implemented.  But I can make some tests locally later.

Reply:
@markuman thank you for the feedback!

To me lines 396-399 say:
1. If the rule does not exist:
2.     If any of the parameters are passed -> Update the rule (why?)
3.     IF not passed -> just create the rule

Step number 2 in conjunction with 1 looks a bit strange to me.
The module makes the `config_data` attr from module.params.
In other words, 1 + 2 say: "If the rule does not exist and if any of the parameters are passed, update the rule"

Shouldn't we `create` the rule if it does not exist and `update` it when it already exists?

And about `check_mode`. The `update` function is:
```
    def update_rule(self, check_mode, result, cursor):
        if not check_mode:
            result['changed'] = self.update_rule_config(cursor)
            result['msg'] = "Updated rule in mysql_query_rules_fast_routing."
            self.manage_config(cursor, result['changed'])
            result['rules'] = self.get_rule_config(cursor)
        else:
            result['changed'] = True
            result['msg'] = (
                "Rule would have been updated in "
                "mysql_query_rules_fast_routing, "
                "however check_mode is enabled."
            )
```
As we can see, if `check_mode=True`, it doesn't compare anything and just returns `True`.
I don't see any comparison of the current state and the desired state.
Maybe it's implemented somehow via the `check_rule_cfg_exists` method, i.e. if the query it constructs returns nothing, consider that the rule does not exist.

@markuman anyway, if you're a ProxySQL user, you can test it locally, and provide feedback that everything works as other ProxySQL users will expect, it would be nice:)